sklearn/gaussian_process/gpr.py

```diff
@@ -0,0 +1,148 @@
+    ----------
+    X_fit_:
+
+    y_fit_:
+
+    theta_:
+
+    L_:
+
+    alpha_:
+    """
+
+    def __init__(self, kernel, y_err=1e-10):
+        self.kernel = kernel
+        self.y_err = y_err
+
+    def fit(self, X, y):
+        self.X_fit_ = np.asarray(X)
+        self.y_fit_ = np.asarray(y)
+
+        if self.kernel.has_bounds:
+            # Choose hyperparameters based on maximizing the log-marginal
+            # likelihood
+            def obj_func(theta):
+                lml, grad = self.log_marginal_likelihood(theta,
+                                                         eval_gradient=True)
+                return -lml, -grad
+            self.theta_, lml, _ = fmin_l_bfgs_b(obj_func, self.kernel.params,
+                                                bounds=self.kernel.bounds)
+            self.kernel.params = self.theta_
+        else:
+            self.theta_ = self.kernel.params
+
+        # Precompute quantities required for predictions which are independent
+        # of actual query points
+        K = self.kernel.auto(self.X_fit_)
+        K[np.diag_indices_from(K)] += self.y_err
+        self.L_ = cholesky(K, lower=True)  # Line 2
+        self.alpha_ = cho_solve((self.L_, True), self.y_fit_)  # Line 3
+
+        return self
+
+    def predict(self, X, return_std=False, return_cov=False):
+        assert not (return_std and return_cov), \
+            "Not returning standard deviation of predictions when " \
+            "returning full covariance."
+        X = np.asarray(X)
+
+        if not hasattr(self, "X_fit_"):  # Unfitted; predict based on GP prior
+            y_mean = np.zeros(X.shape[0])
+            if return_cov:
+                y_cov = self.kernel.auto(X)
+                return y_mean, y_cov
+            elif return_std:
+                # XXX: Compute y_std more efficiently
+                y_std = np.sqrt(np.diag(self.kernel.auto(X)))
+                return y_mean, y_std
+            else:
+                return y_mean
+        else:  # Predict based on GP posterior
+            K_trans = self.kernel.cross(X, self.X_fit_)
+            y_mean = K_trans.dot(self.alpha_)  # Line 4 (y_mean = f_star)
+            if return_cov:
+                v = cho_solve((self.L_, True), K_trans.T)  # Line 5
+                y_cov = \
+                    self.kernel.auto(X) - K_trans.dot(v)  # Line 6
+                return y_mean, y_cov
+            elif return_std:
+                # XXX: Compute y_std more efficiently
+                v = cho_solve((self.L_, True), K_trans.T)  # Line 5
+                y_cov = \
+                    self.kernel.auto(X) - K_trans.dot(v)  # Line 6
+                y_std = np.sqrt(np.diag(y_cov))
+                return y_mean, y_std
+            else:
+                return y_mean
+
+    def sample(self, X, n_samples=1):
+        y_mean, y_cov = self.predict(X, return_cov=True)
+        y_samples = \
+            np.random.multivariate_normal(y_mean, y_cov, n_samples).T
```

Comment:
use a random number generator: `def sample(self, X, n_samples=1, random_state=0)` and in the body `rng = check_random_state(random_state)` (`sklearn.utils.check_random_state`, I think)